agent/util-scripts/pbench-tool-meister-start

```diff
@@ -151,60 +219,98 @@ class ToolGroup:
+        """
+        try:
+            label = self.labels[host]
+        except KeyError:
+            label = ""
+        return label
+
 
-def wait_for_subs(chan, expected_tms, logger):
-    """wait_for_subs - Wait for the data sink and the proper number of TMs to
-    register, and when they are all registered, return a dictionary of the
-    data sink and tool meister(s) pids.
+def wait_for_tds(chan, logger):
+    """wait_for_tds - Wait for the Tool Data Sink to report back success or
+    failure regarding the Tool Meister environment setup.
     """
-    pids = dict()
-    have_ds = False
-    num_tms = 0
-    for payload in chan:
-        try:
-            json_str = payload["data"].decode("utf-8")
-        except Exception:
-            logger.warning("data payload in message not UTF-8, '%r'", json_str)
-            continue
-        logger.debug("w. channel payload, %r", json_str)
-        try:
-            data = json.loads(json_str)
-        except json.JSONDecodeError:
-            logger.warning("data payload in message not JSON, '%s'", json_str)
-            continue
+    status = ""
+    for data in chan.fetch_json(logger):
         # We expect the payload to look like:
-        #   { "kind": "<ds|tm>",
-        #     "hostname": "<hostname>",
-        #     "pid": "<pid>"
+        #   { "kind": "ds",
+        #     "action": "startup",
+        #     "status": "success|failure"
         #   }
-        # Where 'kind' is either 'ds' (data-sink) or 'tm' (tool-meister),
-        # 'hostname' is the host name on which that entity is running, and
-        # 'pid' is that entity's PID on that host.
         try:
-            new_data = dict(
-                kind=data["kind"], hostname=data["hostname"], pid=data["pid"]
-            )
+            kind = data["kind"]
+            action = data["action"]
+            status = data["status"]
         except KeyError:
             logger.warning("unrecognized data payload in message, '%r'", data)
             continue
         else:
-            if new_data["kind"] == "ds":
-                pids["ds"] = new_data
-                have_ds = True
-            elif new_data["kind"] == "tm":
-                if "tm" not in pids:
-                    pids["tm"] = []
-                pids["tm"].append(new_data)
-                num_tms += 1
-            else:
-                logger.warning("unrecognized 'kind', in data payload '%r'", data)
+            if kind != "ds":
+                logger.warning("unrecognized kind field in message, '%r'", data)
+                continue
+            if action != "startup":
+                logger.warning("unrecognized action field in message, '%r'", data)
                 continue
-        if have_ds and num_tms == expected_tms:
             break
-    return pids
+    return 0 if status == "success" else 1
+
 
+class ReturnCode:
+    """ReturnCode - symbolic return codes for when the main program of
+    pbench-tool-meister-start.
+    """
```

Comment:
It helps; but for it to really be stable and useful across builds, we have to make sure that we never change the value of any of these constants. If we stop reporting an error, we can abandon that value, but we can't re-use that value. (And we shouldn't ever remove "abandoned" constants, either, so that even debugging older code we can use the latest source as a valid reference.) That should probably be part of the doc string (and ideally be tested) if this is really supposed to be a solution.

At worst, though, it's a lot easier to read with symbolic names instead of integers.